### DISCUSSION TITLE
Using Blueprint with Time Series Data

### DISCUSSION THREAD

**Discussion by pedropalb** [abstractqqq/polars_ds_extension#370] [Q&A] [answered]
Hi everyone,

I'm working with a time series DataFrame in a tidy format, with a schema similar to this:
``` python
{
    "id": pl.String,
    "timestamp": pl.Datetime,
    "var_1": pl.Float64,
    "var_2": pl.Int32,
    "target": pl.Boolean,
    # ... and so on
}
```
I'm considering polars-ds to build a reproducible and serializable preprocessing pipeline. My goal is to use a `Blueprint` object to process this data, but I need to perform operations on groups based on the id column.
Here’s a summary of the processing steps I need to implement:
 * Global Scaling: Apply scaling to all `var_*` columns across the entire DataFrame. I believe this is straightforward with `Blueprint`.
 * Grouped Signal Processing: Apply different signal processing functions to each `var_*` column, but this time grouped by `id`. I'm unsure if `Blueprint` supports grouped operations, and this is my main question.
 * Windowing and Aggregation: Create windows using `group_by_dynamic` on the `timestamp` column, also grouped by `id`. Within each window, I need to aggregate the target and collect the `var_*` columns into lists, which are then cast to fixed-size arrays. The final schema would look something like this:
``` python
{
    "id": pl.String,
    "features": pl.Array(pl.Float64, 2 * window_size),
    "target": pl.Int32,
    # ...
}
```

 * Model Input: Convert the final DataFrame into a format suitable for model training, for instance, using `df.collect().to_torch()`.
 
Currently, I have a working implementation using pure Polars. However, I'd like to migrate to polars-ds to make the pipeline clearer, more reproducible for the test set, and easily serializable.

Is it possible to achieve the grouped processing (specifically step 2) using a `Blueprint`? Any guidance on how to best structure this pipeline with polars-ds would be greatly appreciated.

Thanks for the great work on this library!

**Comment by abstractqqq** [comment, marked as answer]
Yes. I talked to quite a few people regarding time series. Although I don't want to be too involved in time series in the library, adding group by + other context will eventually make the blueprint object a way to persist query plans for Polars. So it is worth the effort in all sense. I think I will do this: 

1. I will experiment with adding a group by, agg contexts in blueprints. 
2. I will tag you when I have PRs in this direction, and you can see if this works for you. 

If you have existing polars code that does similar things that will be great and I can use that as reference.

**Reply by pedropalb** [reply]
First, sorry for the late response.
Here is a small example of the procedure I described:

```python
import random

import polars as pl
import numpy as np


df = pl.LazyFrame(
    {
        "id": [1] * 9 + [2] * 15 + [3] * 6,
        "timestamp": list(range(9))
        + list(range(15))
        + list(range(6)),
        "var_1": np.random.rand(30),
        "var_2": np.random.rand(30),
        "target": random.choices([False, True], k=30),
    }
)

df = (
    # Global scaling
    df.with_columns(
        (pl.col("var_1") - pl.col("var_1").mean())
        / pl.col("var_1").std(),
        (pl.col("var_2") - pl.col("var_2").mean())
        / pl.col("var_2").std(),
    )
    # Signal processing by id
    .group_by("id", maintain_order=True)
    .agg(
        "timestamp",
        pl.col("var_1").mul(10),
        pl.col("var_2").truediv(10),
        "target",
    )
    .explode(pl.all().exclude("id"))
    # Windowing
    .group_by_dynamic(
        "timestamp",
        every="3i",
        group_by="id",
        start_by="datapoint",
    )
    .agg(
        pl.concat_list("var_1", "var_2").alias("features"),
        pl.col("target").sum(),
    )
    # Cast features to array of shape (window_size, n_channels)
    .with_columns(
        pl.col("features").cast(pl.Array(pl.Float64, (3, 2)))
    )
    .collect()
)

# Create a PyTorch dataset
dataset = df.to_torch("dataset", features="features", label="target")
```

Note that I "manually" applied global scaling in the first step. In practice, I ended up using sklearn's `StandardScaler` to save the mean and std of the training set. I could have used the blueprint object in that step, but then the blueprint would have only a single operation (`scale`) because I believe that the rest of the pipeline cannot be expressed in a blueprint object.

Please, let me know if you need more information about this.

Thank you!

**Reply by abstractqqq** [reply]
I should have a release candidate with the desired functionality soon :) 

A container for `explode` is the only thing missing rn. 

There are some caveats, e.g. `pl.all()` cannot be serialized properly. So it will be slightly different.

**Reply by pedropalb** [reply]
Great! I really appreciate it.

By the way, I could have probably avoided the `group_by` + `explode` by replacing it with something like `df.with_columns(pl.col("var_1").mul(10).over("id"), ...)`.

About the serialization of `pl.all()`, does `polars.selectors` (e.g. `cs.exclude("id")`) would have the same limitations?

**Reply by abstractqqq** [reply]
Just published v0.10.0.

Check the latest example [here](https://github.com/abstractqqq/polars_ds_extension/blob/v0.10.0/examples/pipeline.ipynb)
